tests/drivers/disk/disk_performance/src/main.c

```diff
@@ -0,0 +1,293 @@
+{
+	int rc;
+	timing_t start_time, end_time;
+	uint64_t cycles, total_ns;
+
+	/* Start the timing system */
+	timing_init();
+	timing_start();
+
+	/* Read block we will overwrite, to back it up. */
+	rc = disk_access_read(disk_pdrv, backup_buf, 0, num_blocks);
+	zassert_equal(rc, 0, "disk read failed");
+
+	/* Initialize write buffer with data */
+	sys_rand_get(test_buf, num_blocks * SECTOR_SIZE);
+
+	total_ns = 0;
+	for (int i = 0; i < SEQ_ITERATIONS; i++) {
+		start_time = timing_counter_get();
+
+		rc = disk_access_write(disk_pdrv, test_buf, 0, num_blocks);
+
+		end_time = timing_counter_get();
+
+		zassert_equal(rc, 0, "disk write failed");
+
+		cycles = timing_cycles_get(&start_time, &end_time);
+		total_ns += timing_cycles_to_ns(cycles);
+	}
+	/* Stop timing system */
+	timing_stop();
+
+	/* Replace block with backup */
+	rc = disk_access_write(disk_pdrv, backup_buf, 0, num_blocks);
+	zassert_equal(rc, 0, "disk write failed");
+	/* Return average time */
+	return (total_ns / SEQ_ITERATIONS);
+}
+
+static void test_sequential_write(void)
+{
+	uint64_t time_ns;
+
+	/* Start with single sector write */
+	time_ns = write_helper(1);
+
+	TC_PRINT("Average write speed over one sector: %llu KiB/s\n",
+		((SECTOR_SIZE * (NSEC_PER_SEC / time_ns))) / 1024);
+
+	/* Now time long sequential write */
+	time_ns = write_helper(SEQ_BLOCK_COUNT);
+
+	TC_PRINT("Average write speed over %d sectors: %llu KiB/s\n",
+		SEQ_BLOCK_COUNT,
+		((BUF_SIZE) * (NSEC_PER_SEC / time_ns)) / 1024);
+}
+
+static void test_random_read(void)
+{
+	timing_t start_time, end_time;
+	uint64_t cycles, total_ns;
+	uint32_t sector;
+
+	/* Build list of sectors to read from. */
+	for (int i = 0; i < RANDOM_ITERATIONS; i++) {
+		/* Get random num until we select a value within sector count */
+		do {
+			sector = sys_rand32_get();
+		} while (sector >= (disk_sector_count - 1));
+		chosen_sectors[i] = sector;
+	}
+
+	/* Start the timing system */
+	timing_init();
+	timing_start();
+
+	start_time = timing_counter_get();
+	for (int i = 0; i < RANDOM_ITERATIONS; i++) {
+		/*
+		 * Note: we don't check return code here,
+		 * we want to do I/O as fast as possible
+		 */
+		disk_access_read(disk_pdrv, test_buf, chosen_sectors[i], 1);
+	}
+	end_time = timing_counter_get();
+	cycles = timing_cycles_get(&start_time, &end_time);
+	total_ns = timing_cycles_to_ns(cycles);
+	/* Stop timing system */
+	timing_stop();
+
+	TC_PRINT("512 Byte IOPS over %d random reads: %llu IOPS\n",
+		RANDOM_ITERATIONS,
+		((uint64_t)(((uint64_t)RANDOM_ITERATIONS)*
+		((uint64_t)NSEC_PER_SEC)))
+		/ total_ns);
+}
+
+
+static void test_random_write(void)
+{
+	timing_t start_time, end_time;
+	uint64_t cycles, total_ns;
+	uint32_t sector;
+	int rc;
+
+	/* Build list of sectors to read from. */
+	for (int i = 0; i < RANDOM_ITERATIONS; i++) {
+		/* Get random num until we select a value within sector count */
+		do {
+			sector = sys_rand32_get();
+		} while (sector >= (disk_sector_count - 1));
```

Comment:
why not use a mod operation on the rand result? It would give you some determinism on the execution to not have to skip values outside of the disk_sector_count range

Reply:
Honestly, I'm not sure why I took this approach, my best guess is I wasn't fully awake when I wrote this. I'll update this section to use modulo or integer division.